### PR TITLE
SM/MM: added support for mmap in the mm uct.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/858/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/285/ for details. (Mellanox internal link)

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
Failures:
```
sm/mm/mm_mmap.c: In function 'uct_mmap_alloc':
sm/mm/mm_mmap.c:65: error: 'MAP_HUGETLB' undeclared (first use in this function)
sm/mm/mm_mmap.c:65: error: (Each undeclared identifier is reported only once
sm/mm/mm_mmap.c:65: error: for each function it appears in.)
sm/mm/mm_mmap.c: In function 'uct_mmap_attach':
sm/mm/mm_mmap.c:133: error: 'MAP_HUGETLB' undeclared (first use in this function)
```

File: src/uct/sm/mm/mm_mmap.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        goto err;
+    }
+
+    /* use the mmid (63 bits uuid) to recreate the file_name for shm_unlink */
+    sprintf(file_name, "/ucx_shared_mr_uuid_%zu", mm_id >> 1);
+    if (shm_unlink(file_name) != 0) {
+        ucs_warn("unable to unlink the shared memory segment");
+        status = UCS_ERR_SHMEM_SEGMENT;
+        goto err_free_file;
+    }
```
Comment (shamisp) [inline src/uct/sm/mm/mm_mmap.c]: Who clean files in the case of system failure (abort on execution)?

Comment (shamisp) [inline src/uct/sm/mm/mm_mmap.c]: What is the directory for file allocation

File: src/uct/sm/mm/mm_iface.c [hunk 2]
```diff
@@ -56,8 +57,11 @@ static ucs_status_t uct_mm_iface_get_address(uct_iface_t *tl_iface,
 static int uct_mm_iface_is_reachable(uct_iface_t *tl_iface,
                                      const struct sockaddr *addr)
 {
+    uct_mm_iface_t *iface = ucs_derived_of(tl_iface, uct_mm_iface_t);
+
     return (addr->sa_family == UCT_AF_PROCESS) &&
-           (((uct_sockaddr_process_t*)addr)->node_guid == ucs_machine_guid());
+           (((uct_sockaddr_process_t*)addr)->node_guid ==
+           (ucs_machine_guid() * ucs_string_to_id(iface->super.pd->component->name)));
```
Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: Please clarify motivation for the change

File: src/ucp/proto/ucp_context.c [hunk 3]
```diff
@@ -111,6 +111,12 @@ static int ucp_is_resource_enabled(uct_tl_resource_desc_t *resource,
         }
     }
 
+    /* Disable the mmap and xpmem 'devices'. ONLY for now - use sysv for mm .
+     * This will be removed after multi-rail is supported */
+    if (!strcmp(resource->dev_name,"mmap") || !strcmp(resource->dev_name, "xpmem")) {
+        device_enabled  = 0;
+    }
+
```
Comment (shamisp) [inline src/ucp/proto/ucp_context.c]: OMG: This is really ugly

**Comment by shamisp** [comment]
My understanding this is posix mmap implementation.

File: src/uct/sm/mm/mm_mmap.c (hunk 1 above)
Comment (alinask) [inline src/uct/sm/mm/mm_mmap.c]: The fd associated with the allocateion is created in /dev/shm by default.

File: src/ucp/proto/ucp_context.c (hunk 3 above)
Comment (alinask) [inline src/ucp/proto/ucp_context.c]: This is a temporary workaround. Any solution to hanlde multiple devices until it is actualy supported in ucp, would be temporary. This is the simplest for now.

File: src/uct/sm/mm/mm_iface.c (hunk 2 above)
Comment (alinask) [inline src/uct/sm/mm/mm_iface.c]: sure, i'll add a comment in the code.
it's for preventing mmap and sysv from having the same address for the proc, and therefore prevent them from connecting one to another.

File: src/ucp/proto/ucp_context.c (hunk 3 above)
Comment (yosefe) [inline src/ucp/proto/ucp_context.c]: Due to lack of multi-transport, this is the least intrusive workaround for now

File: src/uct/sm/mm/mm_mmap.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+       if ((*address_p) != MAP_FAILED) {
+           /* indicate that the memory was mapped without hugepages */
+           uuid &= ~UCT_MM_MMAP_HUGETLB;
+           goto out_ok;
+       }
+
+       ucs_debug("mm failed to allocate %zu bytes without hugetlb %m", *length_p);
+    }
+
+err_shm_unlink:
```
Comment (yosefe) [inline src/uct/sm/mm/mm_mmap.c]: i think it makes more sense to close and then unlink

**Comment by yosefe** [comment]
except for small comment, and the need to detect `MAP_HUGETLB` support during compile time, looks good

**Comment by alinask** [comment]
the coverity error in the ud code is fixed in @alex-mikheev 's PR.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/860/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/287/ for details. (Mellanox internal link)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/862/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/289/ for details. (Mellanox internal link)

**Comment by yosefe** [comment]
:+1:

**Comment by yosefe** [comment]
@shamisp can we merge this?

**Comment by shamisp** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/876/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/877/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/878/ for details.

**Pull request merged**